### PR TITLE
Bitwarden Unified Self-Host project

### PR DESCRIPTION
**Pull request by vgrassia**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR contains all of the work on the new Bitwarden Unified Self-Host project.

## Code changes

* **.dockerignore:** Prevents Docker from copying these files/folders when building images.
* **.github/workflows/build.yml:** Clean up formatting, update runners to `ubuntu-22.04`, change `service_name` to `project_name` where needed, remove `K8S-Proxy` Docker image build, add Docker image build for `self-host` (Unified).
* **.github/workflows/release.yml:** Clean up formatting, update runners to `ubuntu-22.04`, change `service_name` to `project_name` where needed, remote `K8S-Proxy` Docker image release, add Docker image release for `self-host` (Unified), add support for `beta` key in Docker release job.
* **docker-unified/.env.example:** Example environment file for Docker Compose to consume.
* **docker-unified/Dockerfile:** Dockerfile to build new Unified Self-Host Docker image.
* **docker-unified/confd/app-id.conf.tmpl:** Template for `app-id.json` file generation.
* **docker-unified/confd/app-id.toml:** Confd config file for `app-id.json` file generation.
* **docker-unified/confd/nginx-config.conf.tmpl:** Template for `bitwarden.conf` NGINX config file generation.
* **docker-unified/confd/nginx-config.toml:** Confd config file to `bitwarden.conf` NGINX config file generation.
* **docker-unified/docker-compose.yml:** Example Docker Compose file for Bitwarden Unified stack.
* **docker-unified/entrypoint.sh:** Entrypoint for Bitwarden Unified Docker image.
* **docker-unified/global-settings.env:** Default settings for Bitwarden Unified Docker image.
* **docker-unified/nginx/logrotate.sh:** Log rotate script for NGINX.
* **docker-unified/nginx/mime.types:** Mime types for NGINX.
* **docker-unified/nginx/nginx.conf:** NGINX base configuration.
* **docker-unified/nginx/proxy.conf:** Proxy config for NGINX.
* **docker-unified/nginx/security-headers-ssl.conf:** Security Headers SSL for NGINX.
* **docker-unified/nginx/security-headers.conf:** Security Headers for NGINX.
* **docker-unified/supervisord/admin.ini:** Supervisord config file for the `Admin` project.
* **docker-unified/supervisord/api.ini:** Supervisord config file for the `API` project.
* **docker-unified/supervisord/events.ini:** Supervisord config file for the `Events` project.
* **docker-unified/supervisord/icons.ini:** Supervisord config file for the `Icons` project.
* **docker-unified/supervisord/identity.ini:** Supervisord config file for the `Identity` project.
* **docker-unified/supervisord/nginx.ini:** Supervisord config file for the `NGINX` project.
* **docker-unified/supervisord/notifications.ini:** Supervisord config file for the `Notifications` project.
* **docker-unified/supervisord/scim.ini:** Supervisord config file for the `SCIM` project.
* **docker-unified/supervisord/sso.ini:** Supervisord config file for the `SSO` project.
* **docker-unified/supervisord/supervisord.conf:** Supervisord base config file.
* **scripts/build:** Move around projects to be in alphabetical order.
* **scripts/build-docker:** Move around projects to be in alphabetical order, remove `K8S-Proxy` project.
* **src/Admin/Admin.csproj:** Add in required migrations projects.
* **src/Admin/HostedServices/DatabaseMigrationHostedService.cs:** Use new abstracted IDbMigrator
* **src/Admin/Startup.cs:** Register IDbMigrator service
* **src/Admin/packages.lock.json:** Update packages and project version.
* **src/Core/Settings/GlobalSettings.cs:** Add in `LogDirectoryByProject` setting.
* **src/Core/Utilities/IDbMigrator.cs:** Abstracted migrator interface
* **src/Core/Utilities/LoggerFactoryExtensions.cs:** Add in logic for new `LogDirectoryByProject` setting.
* **src/Infrastructure.EntityFramework/EntityFrameworkServiceCollectionExtensions.cs:** Specify assemblies for migrator 
* **src/SharedWeb/Utilities/ServiceCollectionExtensions.cs:** return SupportedDatabaseProviders so that we can register the proper IDbMigrator 
* **util/EfShared/MigrationBuilderExtensions.cs:** Redefine namespace
* **util/Migrator/SqlServerDbMigrator.cs:** Implement DB Up logic for Sql Server migration
* **util/Migrator/packages.lock.json:** Add and update packages for `Migrator`.
* **util/MySqlMigrations/Factories.cs:** Remove dependency on Api project
* **util/MySqlMigrations/Migrations/20220322191314_SelfHostF4E.cs:** react to namespace adjustments
* **util/MySqlMigrations/MySqlDbMigrator.cs:** Implement EF migration
* **util/MySqlMigrations/MySqlMigrations.csproj:** Adjust dependencies, namely removing Api project
* **util/PostgresMigrations/Factories.cs:** Remove dependency on Api project
* **util/PostgresMigrations/Migrations/20220322183505_SelfHostF4E.cs:** react to namespace adjustments
* **util/PostgresMigrations/PostgresDbMigratorDbMigrator.cs:** Implement EF migration
* **util/PostgresMigrations/PostgresMigrations.csproj:** Adjust dependencies, namely removing Api project

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by justindbaur — CHANGES_REQUESTED** [review]
Quick first pass

File: util/Migrator/SqlServerDbMigrator.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+public class SqlServerDbMigrator : IDbMigrator
+{
+    private readonly string _connectionString;
+    private readonly ILogger<SqlServerDbMigrator> _logger;
+    private readonly string _masterConnectionString;
+
+    public SqlServerDbMigrator(GlobalSettings globalSettings, ILogger<SqlServerDbMigrator> logger)
+    {
+        _connectionString = globalSettings.MySql.ConnectionString;
```
[the remark this answers is not in the document]
Comment (kspearrin) [reply util/Migrator/SqlServerDbMigrator.cs:20]: I think so, good catch, though I have seen this work in this branch and I am wondering how that was possible.

File: util/MySqlMigrations/MySqlMigrations.csproj [hunk 2]
```diff
@@ -1,12 +1,13 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <UserSecretsId>9f1cd3e0-70f2-4921-8068-b2538fd7c3f7</UserSecretsId>
+    <LangVersion>10</LangVersion>
```
[the remark this answers is not in the document]
Comment (kspearrin) [reply util/MySqlMigrations/MySqlMigrations.csproj:5]: Looks like something Visual Studio added. Not even sure what this is.

**Review by kspearrin — APPROVED** [review]

File: docker-unified/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,263 @@
+###############################################
+#                 Build stage                 #
+###############################################
+FROM node:16-slim AS node-build
+
+# TODO: Change default branch name before merge into master
+ARG web_branch=master
```
Comment (kspearrin) [inline docker-unified/Dockerfile:7]: @vincentsalucci Do we need to do something different here?

Comment (vgrassia) [reply docker-unified/Dockerfile]: I have updated the Dockerfile to pull the latest Web artifact from GitHub.

File: docker-unified/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,269 @@
+###############################################
+#                 Build stage                 #
+###############################################
+FROM alpine:3.16 AS web-setup
```
Comment (kspearrin) [inline docker-unified/Dockerfile:4]: Why not just latest?

Comment (vgrassia) [reply docker-unified/Dockerfile]: Usually we pin versions to minimize pipelines breaking without us making a change, but we can probably use `latest` here.

**Review by kspearrin — APPROVED** [review]

File: .github/workflows/build-self-host.yml [hunk 5]
```diff
@@ -2,12 +2,181 @@
...
+            docker-username,
+            dct-delegate-2-repo-passphrase,
+            dct-delegate-2-key"
+
+      - name: Log into Docker
+        if: |
+          false
+          && (github.ref == 'refs/heads/master' ||
+            github.ref == 'refs/heads/rc' ||
+            github.ref == 'refs/heads/hotfix-rc')
```
Comment (justindbaur) [inline .github/workflows/build-self-host.yml:93]: Wouldn't this never run? Or is that the point right now? Same question for a couple steps in here.

Comment (vgrassia) [reply .github/workflows/build-self-host.yml:93]: Correct, we aren't ready to push to Docker Hub yet.  The `false` flags will be removed at a later date.

**Review by justindbaur — APPROVED** [review]
Looks great!

**Comment by kspearrin** [comment]
@vgrassia Be sure to pick this to rc-2022.12 please

**Comment by vgrassia** [comment]
> @vgrassia Be sure to pick this to rc-2022.12 please

Done.

**Pull request merged**